elkscmd/man/man8/getty.8
.TH getty 8
.SH NAME
getty \- initialise a terminal for login
.SH SYNOPSIS
.B getty
/dev/device [baud]
.SH DESCRIPTION
Set the default terminal parameters including baud rate and
display a system identification banner
.B (/etc/issue)
to the terminal, followed by a login prompt.
Then read a user name from standard
input and execute
.B login
with that name as argument.
.P
This version supports the following meta-codes in 
.B /etc/issue
if the configuration variable
SUPER_SMALL has NOT been defined:
.TP
.B @@
The '@' character.
.TP
.B @B
The current Baud rate of the serial port, taken from the command line.
.TP
.B @D
The current date in DD-MMM-YYYY format. 25-Dec-1984 would be an example.
.TP
.B @H
The current system hostname. If /etc/hostname is found, the contents
of its first line are used, otherwise "LocalHost" is assumed.
.TP
.B @L
The current connection identifier, taken from the command line.
.TP
.B @S
The current system type, always "ELKS".
.TP
.B @T
The current system time in 24 hour HH:MM:SS format.
.. .TP
.. .B @U
.. The current number of connected users.
.TP
.B @V
The current kernel version. This is currently fixed during compilation.
.P
Also supported are the standard C escape codes.
.SH EXIT STATUS
.TP
.I 1
Exec of /bin/login failed.
.SH FILES
.TP
.B /bin/login
The program used to complete the login process.
.TP
.B /etc/hostname
The system hostname (not yet implemented).
.TP
.B /etc/issue
The pre-login message.
.PD
.SH SEE ALSO
.BR login (8),
.BR init (8),
.BR inittab (5)
.SH AUTHORS
Al Riddoch (ajr@ecs.soton.ac.uk)


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/sys_utils/getty.c
/*
 * elkscmd/sys_utils/getty.c
 *
 * Copyright (C) 1998 Alistair Riddoch <ajr@ecs.soton.ac.uk>
 *
 * Source for the /bin/getty command.
 *
 * usage: /bin/getty /dev/tty?? <speed>
 *
 * This file may be distributed under the terms of the GNU General Public
 * License version 2 or, at your option, any later version.
 *
 **************************************************************************
 *
 * This is a small version of getty for use in the ELKS project. It is not
 * fully functional, and may not be the most efficient implementation for
 * larger systems. It minimises memory usage and code size.
 *
 * Support for \? and @? codes has been added in, supporting the following
 * codes:
 *
 *	\@ = @			@@ = @
 *	\\ = \			@B = Baud Rate.
 *	\0 = ^@ 		@D = Date in dd-mmm-yyy format.
 *	\b = ^H 		@H = System hostname.
 *	\f = ^L 		@L = Line identifier.
 *	\n = ^J 		@S = System Identifier.
 *	\r = ^M 		@T = 24 hour time in HH:MM:SS format.
 *	\s = Space		@U = Users connected.
 *	\t = 8-column tab	@V = Kernel version.
 *
 * Note that @U is not yet implemented, and @V returns a fixed string
 * from the compile-time kernel rather than querying the current kernel.
 * These are all works in progress.
 */
#include <sys/types.h>
#include <sys/stat.h>
#include <ctype.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <string.h>
#include <time.h>
#include <termios.h>
#include <stdarg.h>
#include <errno.h>
#include <paths.h>

#define DEBUG		0	/* set =1 for debug messages*/

/* debug messages go here*/
#define CONSOLE		_PATH_CONSOLE

/* For those requiring a super-small getty, the following define cuts out
 * all of the extra functionality regarding the /etc/issue code sequences.
 */
//#define SUPER_SMALL		/* Disable for super-small binary */

#if DEBUG
#define debug		consolemsg
#else
#define debug(...)
#endif

char *	progname;
char	Buffer[64];
int	ch, col = 0;

void consolemsg(const char *str, ...)
{
	static int consolefd = -1;
	char buf[80];

	if (consolefd < 0)
		consolefd = open(CONSOLE, O_RDWR);

	va_list args;
	va_start(args, str);
	sprintf(buf, "%s: ", progname);
	write(consolefd, buf, strlen(buf));
	vsprintf(buf, str, args);
	write(consolefd, buf, strlen(buf));
	va_end(args);
}


#ifndef SUPER_SMALL
char	Host[256], *Date = 0, *Time = 0;

void host(void) {
    char *ptr;
    int fp = open(_PATH_HOSTNAME,O_RDONLY), sz;

    if (fp) {
	sz = read( fp, Host, 255);
	if (sz >= 0)
	    Host[sz] = '\0';
	else
	    *Host = '\0';
	close(fp);
    }
    for (ptr = Host; isprint(*ptr); ptr++)
	continue;
    while (ptr >= &Host[1] && ptr[-1] == ' ')
	ptr--;
    *ptr = '\0';
    if (!*Host)
	strcpy( Host, "LocalHost" );
}

/*	Before  = "Sun Dec 25 12:34:56 7890"
 *	Columns = "0....:....1....:....2..."
 *	After   = "25-Dec-7890 12:34:56 "
 */

void when(void) {
    char *Result;
    time_t now;
    int n;

    if (!Date) {
	now = time(0);
	Result = ctime(&now);

	Result[0]  = Result[8];
	Result[1]  = Result[9];

	Result[3]  = Result[4];
	Result[4]  = Result[5];
	Result[5]  = Result[6];

	Result[7]  = Result[20];
	Result[8]  = Result[21];
	Result[9]  = Result[22];
	Result[10] = Result[23];

	Result[2]  = Result[6]	 = '-';

	for (n=20; n>12; n--)
	    Result[n] = Result[n-1];

	Result[11] = Result[20] = '\0';

	Date = Result;
	if (*Date < '1')
	    Date++;
	Time = Result + 12;
    }
}
#endif

static void put(unsigned char ch)
{
    col++;
    if (ch == '\r' || ch == '\n')
	col = 0;
    write(STDOUT_FILENO, &ch, 1);
}

static void state(char *s)
{
    write(STDOUT_FILENO, s, strlen(s));
}

static speed_t convert_baudrate(speed_t baudrate)
{
	switch (baudrate) {
	case 50: return B50;
	case 75: return B75;
	case 110: return B110;
	case 134: return B134;
	case 150: return B150;
	case 200: return B200;
	case 300: return B300;
	case 600: return B600;
	case 1200: return B1200;
	case 1800: return B1800;
	case 2400: return B2400;
	case 4800: return B4800;
	case 9600: return B9600;
	case 19200: return B19200;
	case 38400: return B38400;
	case 57600: return B57600;
	case 115200: return B115200;
#ifdef B230400
	case 230400: return B230400;
#endif
#ifdef B460800
	case 460800: return B460800;
#endif
#ifdef B500000
	case 500000: return B500000;
#endif
#ifdef B576000
	case 576000: return B576000;
#endif
#ifdef B921600
	case 921600: return B921600;
#endif
#ifdef B1000000
	case 1000000: return B1000000;
#endif
	}
	return 0;
}


int main(int argc, char **argv)
{
    char *ptr;
    int n, fd;
    speed_t baud = 0;
    struct termios termios;

    progname = argv[0];
    signal(SIGTSTP, SIG_IGN);		/* ignore ^Z stop signal*/

    if (argc < 2 || argc > 3) {
	consolemsg("Usage: %s device [baudrate]\n", argv[0]);
	exit(3);
    }

    if (argc == 2)
	debug("startup args '%s'\n", argv[1]);
    else if (argc == 3) {
	baud = atol(argv[2]);
	debug("startup args '%s' %ld\n", argv[1], baud);
    }

    /* allow execution outside of init*/
    if (getppid() != 1) {
	int tty = open(argv[1], O_RDWR);
	if (tty < 0) {
		consolemsg("cannot open terminal %s\n", argv[1]);
		exit(4);
	}

	debug("redirecting stdio to %s\n", argv[1]);
	close(0); close(1); close(2); /* close inherited stdio */
	if (dup2(tty, 0) != 0 || dup2(tty, 1) != 1 || dup2(tty, 2) != 2) {
		consolemsg("cannot redirect stdio (error %d)\n", errno);
		exit(5);
	}
	close(tty);
    }

    /* setup tty termios state*/
    baud = convert_baudrate(baud);
    if (baud) debug("setting termio baudcode %ld\n", baud);
    if (tcgetattr(STDIN_FILENO, &termios) >= 0) {
        termios.c_lflag |= ISIG | ICANON | ECHO | ECHOE;
        termios.c_lflag &= ~(IEXTEN | ECHOK | NOFLSH | ECHONL);
        termios.c_iflag |= BRKINT | ICRNL;
        termios.c_iflag &= ~(IGNBRK | IGNPAR | PARMRK | INPCK | ISTRIP | INLCR | IGNCR
		| IXON | IXOFF | IXANY);
        termios.c_oflag |= OPOST | ONLCR;
        termios.c_oflag &= ~XTABS;
        if (baud)
            termios.c_cflag = baud;
        termios.c_cflag &= ~PARENB;
        termios.c_cflag |= CS8 | CREAD | HUPCL;
        termios.c_cflag |= CLOCAL;			/* ignore modem control lines*/
        //termios.c_cflag |= CRTSCTS;		/* hw flow control*/
        termios.c_cc[VMIN] = 1;
        termios.c_cc[VTIME] = 0;
        tcsetattr(STDIN_FILENO, TCSAFLUSH, &termios);
    } else debug("tcgetattr(0) failed\n");

    fd = open(_PATH_ISSUE, O_RDONLY);
    if (fd >= 0) {
	put('\n');
#ifdef SUPER_SMALL
	while ((n=read(fd,Buffer,sizeof(Buffer))) > 0)
	    write(1,Buffer,n);
#else
	when();
	host();
	*Buffer = '\0';
	while (read(fd,Buffer,1) > 0) {
	    ch = *Buffer;
	    if (ch == '\\' || ch == '@') {
		Buffer[1] = ch;
		read(fd,Buffer+1,1);
	    }
	    switch (ch) {
		case '\n':
		    put(ch);
		    break;
		case '\\':
		    ch = Buffer[1];
		    switch(ch) {
			case '0':			/* NUL */
			    ch = 0;
			case '\\':
			case '@':
			    put(ch);
			    break;
			case 'b':			/* BS Backspace */
			    put(8);
			    break;
			case 'f':			/* FF Formfeed */
			    put(12);
			    break;
			case 'n':			/* LF Linefeed */
			    put(10);
			    break;
			case 's':			/* SP Space */
			    put(32);
			    break;
			case 't':			/* HT Tab */
			    do {
				put(' ');
			    } while (col & 7);
			    break;
			case 'r':			/* CR Return */
			    ch=13;
			default:			/* Anything else */
			    put('\\');
			    put(ch);
			    break;
		    }
		    break;
		case '@':
		    ch = Buffer[1];
		    switch(ch) {
			case '@':
			    put(ch);
			    break;
			case 'B':			/* Baud Rate */
			    if (argc > 2) {
				state(argv[2]);
				state(" Baud");
			    } else
				state("Terminal");
			    break;
			case 'D':			/* Date */
			    state(Date);
			    break;
			case 'H':			/* Host */
			    state(Host);
			    break;
			case 'L':			/* Line used */
			    ptr = rindex(argv[1],'/');
			    if (ptr == NULL)
				ptr = argv[1];
			    state(ptr);
			    break;
			case 'S':			/* System */
			    state("ELKS");
			    break;
			case 'T':			/* Time */
			    state(Time);
			    break;
#if 0
			case 'U':			/* Users */
			    state("1 user");
			    break;
#endif
#ifdef ELKS_VERSION
			case 'V':			/* Version */
			    state(ELKS_VERSION);
			    break;
#endif
			default:
			    put('@');
			    put(ch);
			    break;
		    }
		    break;
		default:
		    put(ch);
		    break;
	    }
	    *Buffer = '\0';
	}
#endif
	close(fd);
    }


    for (;;) {
	state("login: ");
	errno = 0;
	n=read(STDIN_FILENO,Buffer,sizeof(Buffer)-1);
	if (n < 1) {
	    debug("read fail on stdin, errno %d\n", errno);
	    if (errno == EINTR)
		continue;
	    exit(1);
	}
	Buffer[n] = '\0';
	while (n > 0)
	    if (Buffer[--n] < ' ')
		Buffer[n] = '\0';
	if (*Buffer) {
	    char *nargv[3];

	    debug("calling login: %s\n", Buffer);
	    nargv[0] = _PATH_LOGIN;
	    nargv[1] = Buffer;
	    nargv[2] = NULL;
	    execv(nargv[0], nargv);
	    debug("execv fail\n");
	    exit(2);
	}
    }
}
